src/test/clojure/xtdb/kafka_test.clj

```diff
@@ -89,3 +96,207 @@
+                   (set (xt/q node "SELECT _id FROM xt_docs")))))
+
+        (t/testing "can continue to index/query new transactions"
+          (t/is (xt/execute-tx node [[:put-docs :xt_docs {:xt/id :new3}]]))
+          (t/is (= (set [{:xt/id :foo}
+                         {:xt/id :bar} 
+                         {:xt/id :new}
+                         {:xt/id :new2}
+                         {:xt/id :new3}])
+                   (set (xt/q node "SELECT _id FROM xt_docs")))))))))
+
+(defn copy-first-n-messages!
+  [^String bootstrap-servers ^String source-topic ^String dest-topic ^long n]
+  ;; 1. Create the destination topic (always, no check)
+  (with-open [admin (AdminClient/create {"bootstrap.servers" bootstrap-servers})]
+    (.createTopics admin
+                   [(doto (NewTopic. dest-topic 1 (short 1))
+                      (.configs {"message.timestamp.type" "LogAppendTime"}))])
+    (Thread/sleep 500)) ;; allow for topic propagation
+
+  ;; 2. Use ByteBuffer serializers, Unit keys, and read_committed
+  (let [consumer (KafkaConsumer.
+                  ^Map {"bootstrap.servers" bootstrap-servers
+                        "enable.auto.commit" "false"
+                        "isolation.level" "read_committed"
+                        "auto.offset.reset" "earliest"
+                        "key.deserializer" "org.apache.kafka.common.serialization.ByteArrayDeserializer"
+                        "key.serializer" "org.apache.kafka.common.serialization.ByteBufferSerializer"
+                        "value.deserializer" "org.apache.kafka.common.serialization.ByteArrayDeserializer"})
+
+        producer (KafkaProducer.
+                  ^Map {"bootstrap.servers" bootstrap-servers
+                        "enable.idempotence" "true"
+                        "acks" "all"
+                        "compression.type" "snappy"
+                        "key.deserializer" "org.apache.kafka.common.serialization.ByteArrayDeserializer"
+                        "key.serializer" "org.apache.kafka.common.serialization.ByteBufferSerializer"
+                        "value.serializer" "org.apache.kafka.common.serialization.ByteBufferSerializer"})]
+
+    ;; 3. Begin copying first N messages
+    (try
+      (let [tp (TopicPartition. source-topic 0)]
+        (.assign consumer [tp])
+        (.seek consumer tp 0)
+
+        (loop [remaining n]
+          (when (pos? remaining)
+            (let [records (.poll consumer (Duration/ofSeconds 1))
+                  recs (iterator-seq (.iterator records))]
+              (doseq [^ConsumerRecord r (take remaining recs)]
+                (.send producer (ProducerRecord. dest-topic (.key r) (ByteBuffer/wrap (.value r)))))
+              (when (< (count recs) remaining)
+                (recur (- remaining (count recs)))))))
+        (.flush producer))
+      (finally
+        (.close consumer)
+        (.close producer)))))
+
+  (t/deftest ^:integration test-stale-log-recovery
```

Comment:
good to test this here - I'd guess that this is probably easier written against the memory/local logs?

might also be easier to test this using the existing XT implementations - e.g. against our `Log` interface - then you can test all three logs using the same test infra?

Reply:
Perhaps instead of the node level for this - could essentially unit test the Log itself?

Reply:
Perhaps do end-to-end with memory log instead? Or create a new log implementation (ie, SimulatedObjectStore type stuff)